drivers/counter/counter_native_posix.c

```diff
@@ -11,33 +11,81 @@
 
 #define DT_COUNTER_LABEL counter0
 #define DRIVER_CONFIG_INFO_FLAGS (COUNTER_CONFIG_INFO_COUNT_UP)
-#define DRIVER_CONFIG_INFO_CHANNELS 1
+#define DRIVER_CONFIG_INFO_CHANNELS CONFIG_COUNTER_NATIVE_POSIX_CHANNELS_NUMBER
 #define COUNTER_NATIVE_POSIX_IRQ_FLAGS (0)
 #define COUNTER_NATIVE_POSIX_IRQ_PRIORITY (2)
 
 #define COUNTER_PERIOD (USEC_PER_SEC / CONFIG_COUNTER_NATIVE_POSIX_FREQUENCY)
 #define TOP_VALUE (UINT_MAX)
 
-static struct counter_alarm_cfg pending_alarm;
-static bool is_alarm_pending;
+struct counter_alarm {
+	struct counter_alarm_cfg config;
+	bool active;
+};
+
+static struct counter_top_cfg top_cfg;
+static struct counter_alarm alarms[DRIVER_CONFIG_INFO_CHANNELS];
 static const struct device *device;
 
+static bool any_alarm_active(void)
+{
+	for (int ch_id = 0; ch_id < DRIVER_CONFIG_INFO_CHANNELS; ch_id++) {
+		if (alarms[ch_id].active) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void schedule_next_irq(void)
+{
+	uint32_t next_value = top_cfg.ticks;
+	uint32_t current_value = hw_counter_get_value();
+
+	for (int ch_id = 0; ch_id < DRIVER_CONFIG_INFO_CHANNELS; ch_id++) {
+		if (alarms[ch_id].active &&
+		    alarms[ch_id].config.ticks >= current_value &&
+		    alarms[ch_id].config.ticks < next_value) {
+			next_value = alarms[ch_id].config.ticks;
+		}
+	}
+
+	hw_counter_set_target(next_value);
+	irq_enable(COUNTER_EVENT_IRQ);
+}
+
 static void counter_isr(const void *arg)
 {
 	ARG_UNUSED(arg);
 	uint32_t current_value = hw_counter_get_value();
 
-	if (is_alarm_pending) {
-		is_alarm_pending = false;
-		pending_alarm.callback(device, 0, current_value,
-				       pending_alarm.user_data);
+	for (int ch_id = 0; ch_id < DRIVER_CONFIG_INFO_CHANNELS; ch_id++) {
+		if (alarms[ch_id].active &&
+		    alarms[ch_id].config.ticks >= current_value) {
```

Comment:
Should it be ` alarms[ch_id].config.ticks <= current_value`?
Still, it does not take into account wrapping. Also note that it maybe be hard to distinguish between something very far in the future (top - 1) and right in the past. Check for `guard_period` in counter.h. It is supposed to be used in that case to protect against setting value too far in the future and allows determining alarms which are in the past.